### PR TITLE
Correct ssl  recv

### PR DESCRIPTION
**Pull request by moritzj29**
fixes esphome compilation issue when ssl fingerprint is enabled
as discussed here: https://github.com/esphome/issues/issues/786

### CODE REVIEW

File: src/DebugPrintMacros.h [hunk 1]
```diff
@@ -27,18 +27,34 @@ inline struct _DEBUG_TIME_STAMP debugTimeStamp(void) {
...
+#endif
+
+#endif
+
 #if defined(DEBUG_ESP_PORT) && !defined(DEBUG_GENERIC)
   #define DEBUG_GENERIC( module, format, ... ) \
     do { \
       struct _DEBUG_TIME_STAMP st = debugTimeStamp(); \
-      DEBUG_ESP_PORT.printf( DEBUG_TIME_STAMP_FMT module " " format, st.whole, st.dec, ##__VA_ARGS__ ); \
+      DEBUG_ESP_PORT_PRINTF( (DEBUG_TIME_STAMP_FMT module " " format), st.whole, st.dec, ##__VA_ARGS__ ); \
```
Comment (OttoWinter) [inline src/DebugPrintMacros.h:50]: Why these changes relating to logging?

I'd prefer them to be filed in a separate PR to make reviewing easier.

File: src/tcp_axtls.c [hunk 2]
```diff
@@ -387,20 +401,31 @@ int tcp_ssl_read(struct tcp_pcb *tcp, struct pbuf *p) {
     } else if(read_bytes > 0){
       if(fd_data->on_data){
         fd_data->on_data(fd_data->arg, tcp, read_buf, read_bytes);
+        // fd_data may have been freed in callback
+        fd_data = tcp_ssl_get(tcp);
+        if(NULL == fd_data)
+          return SSL_CLOSE_NOTIFY;
```
Comment (OttoWinter) [inline src/tcp_axtls.c:407]: Can you comment on where the actual fix is? I see a lot of changed lines, but I'd like to know where the fixed happened so that I can review this accurately.

**Comment by moritzj29** [comment]
I'm sorry, I cannot comment on any of the changes made, since I did not make them. This is a pending PR of the base repository by @mhightower83. Also @jeroenst did work on this.

Merging these changes into ESPAsyncTCP for esphome helped solving my issues described earlier. Shame on me, I have not worked through the code of these libraries. I'm just thankfully using them.

Feel free to close this PR and pull from the original repo as soon as the changes are merged.

File: src/DebugPrintMacros.h (hunk 1 above)
Comment (mhightower83) [reply src/DebugPrintMacros.h:50]: Some changes were needed to generalize the debug print macro a little more so I could get them to work from tcp_axtls.c.

File: src/tcp_axtls.c (hunk 2 above)
Comment (mhightower83) [reply src/tcp_axtls.c:407]: The fix is at line 418 through 420 and 404 through 407.
The problem occurs when the client does a delete on _client from a callback and fd_data becomes invalid.

**Comment by s00500** [comment]
Hi everyone, super nice work! , I just tested this and it works great. I understand the question of separating logging related changes and the actual fix. But I think merging them together makes sense in this case...
Since this fixes general TLS functionality on ESPHome, can this be merged ? Or is the plan to wait for https://github.com/me-no-dev/ESPAsyncTCP/pull/129 ? (Also not ideal for ESPHome)

**Comment by Variour** [comment]
While this is waiting to be merged, can someone tell me **where** (i can figure out the how) to manually replace these files with the modified versions in my local installation (HASS.io AddOn or pip setup)?

**Comment by moritzj29** [comment]
> 
> 
> While this is waiting to be merged, can someone tell me **where** (i can figure out the how) to manually replace these files with the modified versions in my local installation (HASS.io AddOn or pip setup)?

I can't comment on HASS.io but on CLI: When you want to compile & upload your project, the compilation usually fails due these MQTT SSL errors. At this time platformio already created folders within your project directory storing the dependencies and also the platformio.ini file for your project. Within these (hidden) folders there should be the ESPAsyncTCP library. Look for the corresponding files and replace them. Aftwerwards compile and upload. I used platformio directly instead of esphome for this step.
Everytime platformio re-downloads the libraries, the changes are lost.
There are probably more sophisticated methods telling platfomio which libraries to use, but it works.

**Review by OttoWinter — APPROVED** [review]
Thanks!

And sorry for the delay
